### PR TITLE
update scheme to use GroupVersion

### PR DESCRIPTION
**Pull request by deads2k**
Last two commits are unique.  This updates the scheme to use group version.

@liggitt @caesarxuchao

### CODE REVIEW

**Comment by deads2k** [comment]
@pmorie ref https://github.com/kubernetes/kubernetes/issues/17216

The idea is to pick off small units and handle ripples one layer out.  Conversions happen on the edges

**Comment by deads2k** [comment]
@DirectXMan12 This is the next pull in the chain.  The rest already have lgtm (or permission to tag on prereq merges)

File: pkg/api/testapi/testapi.go [hunk 1]
```diff
@@ -87,6 +88,14 @@ func (g TestGroup) GroupAndVersion() string {
 	return g.GroupVersionUnderTest
 }
 
+func (g TestGroup) InternalGroupVersion() unversioned.GroupVersion {
```
Comment (derekwaynecarr) [inline pkg/api/testapi/testapi.go]: Nit godoc , maybe that contrasts this with the method below.

File: pkg/api/testapi/testapi.go [hunk 2]
```diff
@@ -87,6 +88,14 @@ func (g TestGroup) GroupAndVersion() string {
...
+	return unversioned.GroupVersion{Group: g.Group}
+}
+
+func (g TestGroup) GroupVersion() unversioned.GroupVersion {
```
Comment (derekwaynecarr) [inline pkg/api/testapi/testapi.go]: nit Godoc

File: pkg/api/unversioned/group_version.go [hunk 3]
```diff
@@ -48,9 +48,18 @@ type GroupVersion struct {
 	Version string
 }
 
+func (gv GroupVersion) IsEmpty() bool {
```
Comment (derekwaynecarr) [inline pkg/api/unversioned/group_version.go]: nit: godoc

File: pkg/api/unversioned/group_version.go [hunk 4]
```diff
@@ -84,6 +98,10 @@ func ParseGroupVersionOrDie(gv string) GroupVersion {
 	return ret
 }
 
+func (gv GroupVersion) WithKind(kind string) GroupVersionKind {
```
Comment (derekwaynecarr) [inline pkg/api/unversioned/group_version.go]: nit godoc

File: pkg/conversion/decode.go [hunk 5]
```diff
@@ -21,20 +21,33 @@ import (
...
-	if version == "" && s.InternalVersion != "" {
+
+	if len(gv.Version) == 0 && len(internalGV.Version) != 0 {
 		return nil, "", "", fmt.Errorf("version not set in '%s'", string(data))
 	}
 	if kind == "" {
 		return nil, "", "", fmt.Errorf("kind not set in '%s'", string(data))
 	}
-	obj, err = s.NewObject(version, kind)
+	obj, err := s.NewObject(version, kind)
```
Comment (derekwaynecarr) [inline pkg/conversion/decode.go]: +1 to getting rid of named return arguments ;-)

File: pkg/conversion/scheme.go [hunk 6]
```diff
@@ -19,22 +19,19 @@ package conversion
...
 type Scheme struct {
 	// versionMap allows one to figure out the go type of an object with
 	// the given version and name.
-	versionMap map[string]map[string]reflect.Type
-
-	// typeToVersion allows one to figure out the version for a given go object.
-	// The reflect.Type we index by should *not* be a pointer. If the same type
-	// is registered for multiple versions, the last one wins.
-	typeToVersion map[reflect.Type]string
+	gvkToType map[unversioned.GroupVersionKind]reflect.Type
```
Comment (derekwaynecarr) [inline pkg/conversion/scheme.go]: I would prefer we avoid acronyms if possible:  groupVersionKindToType

If unversioned.GroupVersionKind is too much of a mouthful, we should shorten that into more abstract term ;-)

File: pkg/conversion/scheme.go [hunk 7]
```diff
@@ -19,22 +19,19 @@ package conversion
...
 
-	// typeToKind allows one to figure out the desired "kind" field for a given
-	// go object. Requirements and caveats are the same as typeToVersion.
-	typeToKind map[reflect.Type][]string
+	// typeToGroupVersion allows one to find metadata for a given go object.
+	// The reflect.Type we index by should *not* be a pointer.
+	typeToGVK map[reflect.Type][]unversioned.GroupVersionKind
```
Comment (derekwaynecarr) [inline pkg/conversion/scheme.go]: same general comment here on acronyms.

i think the api server code is hard enough to follow that abbreviating GroupVersionKind makes it worse for new eyes.

also would need to align the name with teh comment.

Comment (derekwaynecarr) [inline pkg/conversion/scheme.go]: I am going to withhold comments on the general usage of 'gvk' throughout the rest of the PR, but want to record my general distaste for abbreviaion, and believe it makes it harder for new folks tor reason about in the long run.

File: pkg/conversion/scheme.go [hunk 8]
```diff
@@ -82,30 +86,32 @@ func (s *Scheme) Log(l DebugLogger) {
...
+		}
+		internalGVK := internalGV.WithKind(gvk.Kind)
+
+		if internalType, exists := s.gvkToType[internalGVK]; exists {
+			return s.typeToGVK[internalType][0].Kind
 		}
 	}
-	return names[0]
+
+	return gvks[0].Kind
```
Comment (derekwaynecarr) [inline pkg/conversion/scheme.go]: any chance we would need to protect for non len(0) above?

File: pkg/runtime/codec.go [hunk 9]
```diff
@@ -75,6 +76,8 @@ type codecWrapper struct {
 	version string
 }
 
+var _ Decoder = &codecWrapper{}
```
Comment (derekwaynecarr) [inline pkg/runtime/codec.go]: nit comment on this is as just an implementation check.

File: pkg/runtime/conversion_generator.go [hunk 10]
```diff
@@ -24,12 +24,13 @@ import (
...
 	"strings"
 
+	"k8s.io/kubernetes/pkg/api/unversioned"
 	"k8s.io/kubernetes/pkg/conversion"
 	"k8s.io/kubernetes/pkg/util/sets"
 )
 
 type ConversionGenerator interface {
-	GenerateConversionsForType(version string, reflection reflect.Type) error
+	GenerateConversionsForType(groupVersion unversioned.GroupVersion, reflection reflect.Type) error
```
Comment (derekwaynecarr) [inline pkg/runtime/conversion_generator.go]: this is an example of where we are inconsistent in this PR.

groupVersion unversioned.GroupVersion

and later:

gv unversioned.GroupVersion

File: pkg/conversion/scheme.go (hunk 8 above)
Comment (deads2k) [inline pkg/conversion/scheme.go]: > any chance we would need to protect for non len(0) above?

No.  If it exists in the map, we have at least one element.

File: pkg/runtime/scheme.go [hunk 11]
```diff
@@ -244,27 +259,31 @@ func NewScheme() *Scheme {
 	return s
 }
 
+func (s *Scheme) AddInternalGroupVersion(gv unversioned.GroupVersion) {
```
Comment (derekwaynecarr) [inline pkg/runtime/scheme.go]: nit: godoc

File: pkg/runtime/scheme.go [hunk 12]
```diff
@@ -476,8 +495,8 @@ func (s *Scheme) DecodeInto(data []byte, obj Object) error {
 	return s.raw.DecodeInto(data, obj)
 }
 
-func (s *Scheme) DecodeIntoWithSpecifiedVersionKind(data []byte, obj Object, version, kind string) error {
-	return s.raw.DecodeIntoWithSpecifiedVersionKind(data, obj, version, kind)
+func (s *Scheme) DecodeIntoWithSpecifiedVersionKind(data []byte, obj Object, gvk unversioned.GroupVersionKind) error {
```
Comment (derekwaynecarr) [inline pkg/runtime/scheme.go]: nit godoc: (though it wasnt there before, looks like most of these file has it)

**Comment by derekwaynecarr** [comment]
usage of abbreviations aside, i like what this PR does, fix godoc nits, then LGTM

File: pkg/runtime/conversion_generator.go (hunk 10 above)
Comment (DirectXMan12) [inline pkg/runtime/conversion_generator.go]: I think in field names and parameters, it should be spelled out.  I'm personally less concerned about usage within the actual functions.

File: pkg/conversion/scheme.go [hunk 13]
```diff
@@ -388,13 +399,13 @@ func (s *Scheme) ObjectVersionAndKind(obj interface{}) (apiVersion, kind string,
...
-	kinds, kOK := s.typeToKind[t]
-	if !vOK || !kOK {
+
+	gvks, ok := s.typeToGVK[t]
+	if !ok {
 		return "", "", &notRegisteredErr{t: t}
 	}
-	apiVersion = version
-	kind = kinds[0]
+	apiVersion = gvks[0].GroupVersion().String()
```
Comment (DirectXMan12) [inline pkg/conversion/scheme.go]: Should this method signature be changed to return a GroupVersion instead of a string?

Comment (deads2k) [inline pkg/conversion/scheme.go]: > Should this method signature be changed to return a GroupVersion instead of a string?

Yes, but I have to limit ripples.  That forces updates in another 300 spots.

**Comment by deads2k** [comment]
I will update godoc and add the abbreviation gripe to the overall issue to fix as a followup rather than re-rippling these changes.

I really hope the kube internal e2e settles down.

File: pkg/conversion/decode.go [hunk 14]
```diff
@@ -100,7 +130,7 @@ func (s *Scheme) DecodeInto(data []byte, obj interface{}) error {
 // The function then implements the functionality of DecodeInto.
 // If specifiedVersion and specifiedKind are empty, the function degenerates to
 // DecodeInto.
-func (s *Scheme) DecodeIntoWithSpecifiedVersionKind(data []byte, obj interface{}, specifiedVersion, specifiedKind string) error {
+func (s *Scheme) DecodeIntoWithSpecifiedVersionKind(data []byte, obj interface{}, gvk unversioned.GroupVersionKind) error {
```
Comment (caesarxuchao) [inline pkg/conversion/decode.go]: Could you change the name of gvk to specifiedGVK and change the comment accordingly? Or leave a note in the comment saying the gvk contains the "specified" group/version/kind?

File: pkg/api/unversioned/group_version.go [hunk 15]
```diff
@@ -62,6 +71,11 @@ func (gv GroupVersion) String() string {
 // ParseGroupVersion turns "group/version" string into a GroupVersion struct. It reports error
 // if it cannot parse the string.
 func ParseGroupVersion(gv string) (GroupVersion, error) {
+	// this can be the internal version for kube
+	if (len(gv) == 0) || (gv == "/") {
```
Comment (caesarxuchao) [inline pkg/api/unversioned/group_version.go]: The comment is wrong, it not for kube, it's for v1.

So both "" and "/" will be interpreted as the internal version for v1? How about we only allow ""?

File: pkg/api/unversioned/group_version.go [hunk 16]
```diff
@@ -48,9 +48,18 @@ type GroupVersion struct {
...
 }
 
+func (gv GroupVersion) IsEmpty() bool {
+	return len(gv.Group) == 0 && len(gv.Version) == 0
+}
+
 // String puts "group" and "version" into a single "group/version" string. For the legacy v1
 // it returns "v1".
 func (gv GroupVersion) String() string {
+	// special case the internal apiVersion for kube
```
Comment (caesarxuchao) [inline pkg/api/unversioned/group_version.go]: Same here. It's not for kube, it's for the legacy v1.

Comment (caesarxuchao) [inline pkg/api/unversioned/group_version.go]: So the internal version of other groups will be "some-group/"? Why not making them "some-group", so it's consistent with using "" for "v1"?
I'm sorry if this is already discussed somewhere else.

Comment (liggitt) [inline pkg/api/unversioned/group_version.go]: I think the "foo/" string serialization is temporary until these refactors are complete. The internal version should never be serialized to a string in my opinion. "foo/" let's us split, get two tokens, and avoid more special cases

Comment (caesarxuchao) [inline pkg/api/unversioned/group_version.go]: Yes, that would just be temporary. Thank you.

File: pkg/apiserver/resthandler.go [hunk 17]
```diff
@@ -684,15 +686,29 @@ func DeleteResource(r rest.GracefulDeleter, checkBody bool, scope RequestScope,
 // to use it.
 // TODO: add appropriate structured error responses
 func queryToObject(query url.Values, scope RequestScope, kind string) (runtime.Object, error) {
```
Comment (caesarxuchao) [inline pkg/apiserver/resthandler.go]: FYI, we are removing queryToObject here #17163.

Comment (caesarxuchao) [inline pkg/apiserver/resthandler.go]: @deads2k #17163 it's is using the ServerAPIVersion https://github.com/kubernetes/kubernetes/pull/17163/files#diff-fe6dbe5fa548ea5c031a5be1669a6376R242

Comment (liggitt) [inline pkg/apiserver/resthandler.go]: I think ServerAPIVersion is wrong. Groups should be using query options registered to their own group/version, or unversioned query options, but not ones from v1

Comment (liggitt) [inline pkg/apiserver/resthandler.go]: I'd rather see us remove ServerGroupVersion and follow up by cleaning up the "v1" fallback here than expanding the use of ServerGroupVersion like #17163 is

File: pkg/api/unversioned/group_version.go (hunk 15 above)
Comment (caesarxuchao) [inline pkg/api/unversioned/group_version.go]: I think this is also temporary so never mind.

File: pkg/conversion/decode.go (hunk 14 above)
Comment (caesarxuchao) [inline pkg/conversion/decode.go]: I'm ok if you do it in another PR.

**Comment by caesarxuchao** [comment]
@k8s-bot unit test this

**Comment by caesarxuchao** [comment]
I've done my pass and SGTM (I'm not confident to say LGTM with just one pass). Thank you all!

**Comment by deads2k** [comment]
@derekwaynecarr comments addressed, rebased, and one commit added for test-cmd to pass.

**Comment by derekwaynecarr** [comment]
LGTM

**Comment by derekwaynecarr** [comment]
poke after reduce commits, and will apply tag

**Comment by derekwaynecarr** [comment]
Thanks!

File: pkg/api/meta/restmapper.go [hunk 18]
```diff
@@ -211,9 +211,24 @@ func (m *DefaultRESTMapper) GroupForResource(resource string) (string, error) {
...
+	for _, gvString := range versions {
+		currGroupVersion, err := unversioned.ParseGroupVersion(gvString)
+		if err != nil {
+			return nil, err
+		}
+		if len(currGroupVersion.Version) != 0 {
+			hadVersion = true
+		}
+
+		requestedGroupVersions = append(requestedGroupVersions, currGroupVersion)
```
Comment (caesarxuchao) [inline pkg/api/meta/restmapper.go]: What's the usage of this `requestedGroupVersions`?

**Comment by deads2k** [comment]
@k8s-bot unit test this please

**Comment by deads2k** [comment]
cleaned up dead local variable.  re-applying lgtm

**Comment by wojtek-t** [comment]
@deads2k - please rebase

**Comment by wojtek-t** [comment]
@deads2k - I'm afraid you need to rebase again

**Comment by deads2k** [comment]
> @deads2k - I'm afraid you need to rebase again

yay!

**Comment by deads2k** [comment]
@wojtek-t Any idea why this isn't in the queue?

**Comment by deads2k** [comment]
> @wojtek-t Any idea why this isn't in the queue?

and there it is

**Comment by wojtek-t** [comment]
@deads2k - it wasn't because our continuous e2e were red

Now it's close to the top.

**Comment by deads2k** [comment]
rebased again

**Comment by smarterclayton** [comment]
I didn't catch this earlier but putting GroupVersion and GroupVersionKind in api/unversioned (and using them in pkg/runtime interfaces) is really the wrong way around.  Generic interfaces like that belong in conversion or runtime.  The struct returned by api/unversioned and the struct used by the conversion interfaces really have nothing to do with each other.

**Comment by caesarxuchao** [comment]
@smarterclayton, could you develop the point? Are you worrying about the import tree being too large?

**Comment by smarterclayton** [comment]
More that unversioned is not something that should be a dependency of conversion.  Unversioned is an API group that is a set of things that must exist outside (for now) a versioned API.  Conversion is about object transformation - the GVK in unversioned is about wire serialization.  The GVK in conversion and runtime scheme is about version transformation.  The fact that they look the same is not related.

Pretty soon, unversioned will have a dependency on protobuf.  That is not something conversion or runtime should have a dependency on.  The wire serialization of GVK does not have to match the internal object we use in interface selection - if we change the wire object, we should not have to change the internal interfaces.

Finally, conversion package should not really be aware of API versions.  The process to convert Pod in v1 to Pod in v1beta1 is looking up go types in a type registry (runtime.Scheme) and performing sequential conversions via a convertor.  With deserialization moving out of conversion in my codec split pr, it becomes more obvious how simple conversion can be, and how runtime.Scheme layers on top of it (and how wire encoding is layered on top of that).  Separating the two GVK objects clearly then ensures there is no overlapping concerns.
